src/lib/provable/bigint.ts

```diff
@@ -0,0 +1,957 @@
+     * Cost: Moderate
+     * @param a The ProvableBigInt to compare
+     * @returns A Bool indicating if a is less than or equal to b
+     */
+    lessThanOrEqual(a: ProvableBigInt_): Bool {
+      return this.fields
+        .map((field, i) => ({
+          isLess: field.lessThan(a.fields[i]),
+          isEqual: field.equals(a.fields[i]),
+        }))
+        .reduce((result, { isLess, isEqual }) => isLess.or(result.and(isEqual)), new Bool(false))
+        .or(this.equals(a));
+    }
+
+    /**
+     * Checks if one ProvableBigInt is equal to another
+     * Cost: Cheap
+     * @param a The ProvableBigInt to compare
+     * @returns A Bool indicating if a is equal to b
+     */
+    equals(a: ProvableBigInt_): Bool {
+      return this.fields
+        .map((field, i) => field.equals(a.fields[i]))
+        .reduce((result, isEqual) => result.and(isEqual), new Bool(true));
+    }
+
+    /**
+     * Checks if one ProvableBigInt is less than or equal to another
+     * Cost: Cheap
+     * @param a The ProvableBigInt to compare
+     * @returns A Bool indicating if a is less than or equal to b
+     */
+    assertEquals(a: ProvableBigInt_) {
+      this.equals(a).assertTrue('ProvableBigInts are not equal');
+    }
+  }
+  return ProvableBigInt_;
+}
+
+abstract class ProvableBigInt<T> {
+  fields: Field[];
+  value: Unconstrained<bigint>;
+
+  static _provable?: ProvablePureExtended<
+    ProvableBigInt<ProvableBigInt<any>>,
+    { fields: bigint[] },
+    { fields: string[] }
+  >;
+
+  static get provable() {
+    assert(this._provable !== undefined, 'ProvableBigInt not initialized');
+    return this._provable;
+  }
+
+  public static _modulus?: ProvableBigInt<any>;
+  public static _config?: BigIntParameter;
+
+  static get modulus(): ProvableBigInt<any> {
+    assert(this._modulus !== undefined, 'Modulus not initialized');
+    return this._modulus;
+  }
+
+  static get config(): BigIntParameter {
+    assert(this._config !== undefined, 'Config not initialized');
+    return this._config;
+  }
+
+  constructor(fields: Field[], value: Unconstrained<bigint>) {
+    this.fields = fields;
+    this.value = Unconstrained.from(value);
+  }
+
+  abstract get Constructor(): typeof ProvableBigInt;
+  abstract toBigInt(): bigint;
+  abstract toFields(): Field[];
+  abstract toBits(): Bool[];
+  abstract clone(): T;
+  abstract add(a: T, isDouble?: boolean): T;
+  abstract double(): T;
+  abstract sub(a: T): T;
+  abstract mul(a: T, isSquare?: boolean): T;
+  abstract square(): T;
+  abstract div(a: T): T;
+  abstract inverse(): T;
+  abstract negate(): T;
+  abstract pow(exp: T): T;
+  abstract sqrt(): T;
+  abstract mod(): T;
+  abstract greaterThan(a: T): Bool;
+  abstract greaterThanOrEqual(a: T): Bool;
+  abstract lessThan(a: T): Bool;
+  abstract lessThanOrEqual(a: T): Bool;
+  abstract equals(a: T): Bool;
+  abstract assertEquals(a: T): void;
+}
+
+function rangeCheck(x: Field, bits: number, signed?: boolean) {
```

Comment:
imo we should move these into `range-check.ts`

Reply:
what about `rangeCheck48` and co that come beneath this function? For this particular function, maybe a better name would be `rangeCheckSelector` or `rangeCheckHelper`

Reply:
IMHO this method is a classical example of premature abstraction and doesn't belong in `range-check.ts`
* makes the code less readable, e.g. `rangeCheck(x, 128, true)` is less clear than `rangeCheck128Signed(x)`
* adds one additional layer to go from where it's used to where the logic is => harder to navigate
* doesn't save any lines of code (actually it adds code, namely this function)

I would remove it and inline the different ways of doing range checks in this file!